apm-agent-core/src/main/java/co/elastic/apm/agent/db/signature/SignatureParser.java

```diff
@@ -51,15 +54,35 @@
      * We don't want to keep alive references to huge query strings
      */
     private static final int QUERY_LENGTH_CACHE_UPPER_THRESHOLD = 10_000;
+
+    private final Allocator<Scanner> scannerAllocator;
+    private final DetachedThreadLocal<Scanner> SCANNER = new DetachedThreadLocal<Scanner>(DetachedThreadLocal.Cleaner.INLINE) {
+        @Override
+        protected Scanner initialValue(Thread thread) {
+            return scannerAllocator.createInstance();
+        }
+    };
+
     /**
      * Not using weak keys because ORMs like Hibernate generate equal SQL strings for the same query but don't reuse the same string instance.
      * When relying on weak keys, we would not leverage any caching benefits if the query string is collected.
      * That means that we are leaking Strings but as the size of the map is limited that should not be an issue.
      */
-    private final static ConcurrentMap<String, String[]> signatureCache = new ConcurrentHashMap<String, String[]>(DISABLE_CACHE_THRESHOLD,
+    private final ConcurrentMap<String, String[]> signatureCache = new ConcurrentHashMap<String, String[]>(DISABLE_CACHE_THRESHOLD,
             0.5f, Runtime.getRuntime().availableProcessors());
 
-    private final Scanner scanner = new Scanner();
+    public SignatureParser() {
+        this(new Allocator<Scanner>() {
```

Comment:
[minor - recommendation only] the usage of `Allocator` implies object pooling, which is a bit misleading here. Not very important, but since this is a minimal interface, duplicating it for general purposes is fine. Another option is to extract it to another, non-objectpooling package (and adjusting its javadoc).